### ISSUE TITLE
grad with lax.scan causing slow compilation

### ISSUE THREAD

**Issue by edfong** [google/jax#2609]
This is my first issue post, so sorry if I say anything dumb! I'm really enjoying using Jax, it suits my current project perfectly and the performance is great. 

I'm not sure if this is a bug, but I've encountered prohibitively slow compilation when reverse-mode differentiating through a lax.scan, compared to jacfwd. The run-times _after_ compilation are fast and make sense though (roughly equal between jacfwd and jacrev), so I'm not sure if it's an issue with XLA. Could it be an issue with ndtri?

I've tried to illustrate it with a simplified self-contained version of my code below, but it's much worse in my actual code.

```python
import jax.numpy as np
from jax import grad, jit, vmap,jacfwd,jacrev
from jax.scipy.stats import norm
from jax.scipy.special import ndtri,erfc
from jax.lax import scan

@jit #Calculate normal copula cdf
def norm_copula_distribution(u,v,rho):
    pu = ndtri(u)
    pv = ndtri(v)
    z = (pu - rho*pv)/np.sqrt(1- rho**2)
    cop_dist = erfc(-z / np.sqrt(2)) / 2
    return cop_dist

@jit
def lax_f(carry,rng):
    cop_dist,rho = carry
    u = cop_dist
    v = 0.9*cop_dist
    cop_dist = norm_copula_distribution(u,v,rho)
    return (cop_dist,rho),rng

@jit
def lax_f_scan(carry,rng):
    return scan(lax_f,carry,rng)

@jit
def test(cop_dist,rho):
    rng = np.array([0])
    (cop_dist,rho),rng = lax_f_scan((cop_dist,rho),rng)
    
    #this is much faster to compile
    #for i in range(1): 
    #    (cop_dist,rho),_ = lax_f((cop_dist,rho),i)
    return cop_dist

grad_rev = jit(jacrev(test,(1)))
grad_fwd = jit(jacfwd(test,(1)))
```

With lax.scan, the compilation (first run) time for jacrev is much slower, shown below. It's even more noticeably in my main code (18s for jacfwd vs 95s for jacrev for one example).

```python
cop_dist =0.7
rho = 0.95

%time print(test(cop_dist,rho).block_until_ready()) 299 ms
%time print(grad_fwd(cop_dist,rho).block_until_ready()) 1.07 s
%time print(grad_rev(cop_dist,rho).block_until_ready()) 5.76 s
```

If I use a regular Python loop instead, commented out above, the compilation times are reasonable and equal between the two modes:
```python
cop_dist =0.7
rho = 0.95

%time print(test(cop_dist,rho).block_until_ready()) 157 ms
%time print(grad_fwd(cop_dist,rho).block_until_ready()) 185 ms
%time print(grad_rev(cop_dist,rho).block_until_ready()) 198 ms
```

These results are on CPU btw. I've tried on a GPU and it's not as bad but still noticeably slower.

Would really appreciate some help with this! Thanks :)

**Comment by edfong** [comment]
I've seemed to fix this by defining the reverse-mode gradient of ndtri directly: 

```python
from jax import custom_vjp
@custom_vjp
def ndtri_(u):
    return ndtri(u)
def f_fwd(u):
    fu = ndtri_(u)
    return fu,1/norm.pdf(fu)
def f_bwd(gradu,u_bar):
    return (gradu*u_bar,)
ndtri_.defvjp(f_fwd,f_bwd)

%time print(test(cop_dist,rho).block_until_ready()) 293 ms
%time print(grad_rev(cop_dist,rho).block_until_ready()) 538 ms
```

Could the slow down be due to the way lax.scan/grad handle the approximation of ndtri?

**Comment by jeffgortmaker** [comment]
I also have an application that uses `scan` and `ndtri`. I can confirm that using your custom gradient for `ndtri` helps a lot in my application too! My guess is that the analytic gradient just hasn't been added to JAX yet.
